src/restrictedProp.js

```diff
@@ -1,12 +1,18 @@
 import wrapValidator from './helpers/wrapValidator';
 
-function restrictedProp(props, propName, componentName) {
-  if (props[propName] != null) {
-    return new TypeError(`The ${propName} prop on ${componentName} is not allowed.`);
-  }
+function customMessageWrapper(messsageFunction) {
+  function restrictedProp(props, propName, componentName, location) {
+    if (messsageFunction && typeof messsageFunction === 'function') {
+      return messsageFunction(props, propName, componentName, location);
+    }
+    if (props[propName] != null) {
+      return new TypeError(`The ${propName} prop on ${componentName} is not allowed.`);
```

Comment:
Let's replace "prop" here with `${location}` (and make sure to update the tests to pass an explicit location value)